e2e manifests

pohly opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

/kind cleanup

**What this PR does / why we need it**:

For CSI testing we want to get away from having to replicate .yaml files in code. Now the CSI drivers and their RBAC rules get created from .yaml files.

**Special notes for your reviewer**:

My proposal is to review and potentially merge this PR first, then create the `rbac.yaml` in the individual repos at the URLs linked to here in the readmes.
 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
- The builtin system:csi-external-provisioner and system:csi-external-attacher cluster roles
  are deprecated and will not be updated for deployments of CSI sidecar container versions >= 0.4.
  Deployments with the current CSI sidecar containers have to provide their own RBAC
  definitions. The reason is that the rules depend on how the sidecar containers are used,
  which is defined by the deployment.
```

/sig storage
/sig testing
/cc @msau42

pohly committed:
e2e: update bazel BUILD files

Generated via hack/update-bazel.sh.

pohly committed:
e2e framework: UniqueName

This is needed for tests which have to create entities that aren't
namespaced, like ServiceAccounts.

pohly committed:
e2e: deploy from manifest files + enhance CSI volume output

Ensuring that CSI drivers get deployed for testing exactly as intended
was problematic because the original .yaml files had to be converted
into code. e2e/manifest helped a bit, but not enough:
- could not load all entities
- didn't handle loading .yaml files with multiple entities
- actually creating and deleting entities still had to be done in tests

The new framework utility code handles all of that, including the
tricky cleanup operation that tests got wrong (AfterEach does not get
called after test failures!).

In addition, it is ensuring that each test gets its own instance of the
entities.

The PSP role binding for hostpath is now necessary because we switch
from creating a pod directly to creation via the StatefulSet
controller, which runs with less privileges.

Without this, the hostpath test runs into these errors in the
kubernetes-e2e-gce job:

Oct 19 16:30:09.225: INFO: At 2018-10-19 16:25:07 +0000 UTC - event for csi-hostpath-attacher: {statefulset-controller } FailedCreate: create Pod csi-hostpath-attacher-0 in StatefulSet csi-hostpath-attacher failed error: pods "csi-hostpath-attacher-0" is forbidden: unable to validate against any pod security policy: []
Oct 19 16:30:09.225: INFO: At 2018-10-19 16:25:07 +0000 UTC - event for csi-hostpath-provisioner: {statefulset-controller } FailedCreate: create Pod csi-hostpath-provisioner-0 in StatefulSet csi-hostpath-provisioner failed error: pods "csi-hostpath-provisioner-0" is forbidden: unable to validate against any pod security policy: []
Oct 19 16:30:09.225: INFO: At 2018-10-19 16:25:07 +0000 UTC - event for csi-hostpathplugin: {daemonset-controller } FailedCreate: Error creating: pods "csi-hostpathplugin-" is forbidden: unable to validate against any pod security policy: []

The extra role binding is silently ignored on clusters which don't
have this particular role.

pohly commented:
@msau42 I went ahead and also rewrote the GCE CSI test, which allowed me to remove a lot of code from `csi_objects.go`. Can you perhaps check that I didn't break that test? Or is it running in the Kubernetes CI?

pohly commented:
Note that the `csiTestDriver` implementation for hostpath and GCE now use almost the same code. The real difference is in the set of .yaml files and in the names of the containers. It seems very feasible now to just have one struct which defines these settings and then fill in that struct from .yaml files, either shipped with the test suite or provided by a user.

Only `createGCESecrets()` may still need some special handling.

davidz627 commented:
/cc

pohly commented:
Some help would be welcome. The gcePD test failed in https://prow.k8s.io/log?job=pull-kubernetes-e2e-gce&id=56220 again because of permission issues. I don't understand what the error messages are trying to tell me, the service account setup looks sane to me:

```
I1017 18:50:42.646] Oct 17 18:45:34.524: INFO: creating ServiceAccount:
I1017 18:50:42.646] {
I1017 18:50:42.646]   "kind": "ServiceAccount",
I1017 18:50:42.646]   "apiVersion": "v1",
I1017 18:50:42.646]   "metadata": {
I1017 18:50:42.646]     "name": "csi-driver-registrar",
I1017 18:50:42.646]     "namespace": "e2e-tests-csi-volumes-nh7pq",
I1017 18:50:42.646]     "creationTimestamp": null
I1017 18:50:42.646]   }
I1017 18:50:42.646] }
...
I1017 18:50:42.651] Oct 17 18:45:34.531: INFO: creating ClusterRole
I1017 18:50:42.651] :{
I1017 18:50:42.651]   "kind": "ClusterRole",
I1017 18:50:42.651]   "apiVersion": "rbac.authorization.k8s.io/v1",
I1017 18:50:42.651]   "metadata": {
I1017 18:50:42.651]     "name": "driver-registrar-runner-e2e-tests-csi-volumes-nh7pq",
I1017 18:50:42.651]     "creationTimestamp": null
I1017 18:50:42.652]   },
I1017 18:50:42.652]   "rules": [
I1017 18:50:42.652]     {
I1017 18:50:42.652]       "verbs": [
I1017 18:50:42.652]         "get",
I1017 18:50:42.652]         "list",
I1017 18:50:42.652]         "watch",
I1017 18:50:42.652]         "create",
I1017 18:50:42.652]         "update",
I1017 18:50:42.652]         "patch"
I1017 18:50:42.652]       ],
I1017 18:50:42.653]       "apiGroups": [
I1017 18:50:42.653]         ""
I1017 18:50:42.653]       ],
I1017 18:50:42.653]       "resources": [
I1017 18:50:42.653]         "events"
I1017 18:50:42.653]       ]
I1017 18:50:42.653]     },
I1017 18:50:42.653]     {
I1017 18:50:42.653]       "verbs": [
I1017 18:50:42.653]         "get",
I1017 18:50:42.654]         "update",
I1017 18:50:42.654]         "patch"
I1017 18:50:42.654]       ],
I1017 18:50:42.654]       "apiGroups": [
I1017 18:50:42.654]         ""
I1017 18:50:42.654]       ],
I1017 18:50:42.654]       "resources": [
I1017 18:50:42.654]         "nodes"
I1017 18:50:42.654]       ]
I1017 18:50:42.654]     }
I1017 18:50:42.654]   ]
I1017 18:50:42.655] }
...
I1017 18:50:42.657] Oct 17 18:45:34.539: INFO: creating ClusterRoleBinding:
I1017 18:50:42.657] {
I1017 18:50:42.657]   "kind": "ClusterRoleBinding",
I1017 18:50:42.657]   "apiVersion": "rbac.authorization.k8s.io/v1",
I1017 18:50:42.657]   "metadata": {
I1017 18:50:42.657]     "name": "csi-driver-registrar-role-e2e-tests-csi-volumes-nh7pq",
I1017 18:50:42.658]     "creationTimestamp": null
I1017 18:50:42.658]   },
I1017 18:50:42.658]   "subjects": [
I1017 18:50:42.658]     {
I1017 18:50:42.658]       "kind": "ServiceAccount",
I1017 18:50:42.658]       "name": "csi-driver-registrar",
I1017 18:50:42.658]       "namespace": "e2e-tests-csi-volumes-nh7pq"
I1017 18:50:42.658]     }
I1017 18:50:42.658]   ],
I1017 18:50:42.658]   "roleRef": {
I1017 18:50:42.659]     "apiGroup": "rbac.authorization.k8s.io",
I1017 18:50:42.659]     "kind": "ClusterRole",
I1017 18:50:42.659]     "name": "driver-registrar-runner-e2e-tests-csi-volumes-nh7pq"
I1017 18:50:42.659]   }
I1017 18:50:42.659] }
...
I1017 18:45:34.534] Oct 17 18:45:27.723: INFO: At 2018-10-17 18:40:25 +0000 UTC - event for csi-gce-node: {daemonset-controller } FailedCreate: Error creating: pods "csi-gce-node-" is forbidden: unable to validate against any pod security policy: []
I1017 18:45:34.535] Oct 17 18:45:27.723: INFO: At 2018-10-17 18:40:26 +0000 UTC - event for csi-gce-controller: {statefulset-controller } FailedCreate: create Pod csi-gce-controller-0 in StatefulSet csi-gce-controller failed error: pods "csi-gce-controller-0" is forbidden: unable to validate against any pod security policy: []
```

@liggitt @davidz627

liggitt commented:
the current use of that manifest appears to require elevated PSP permissions, set up here:

https://github.com/kubernetes/kubernetes/blob/30a06af453059ffe636f483b7db914d51757139a/test/e2e/storage/csi_volumes.go#L402-L403

are those permissions still getting set up in the namespace where the CSI driver is getting deployed?

davidz627 commented:
@pohly, the CI runs in an environment with PodSecurityPolicy on, you can emulate this in local testing with: `ENABLE_POD_SECURITY_POLICY=true`

Since the GCE PD Daemonset + Statefulset don't use the default test service account, you have to give the elevated test pod security policies to the service accounts linked to the GCE PD Deployments.

@liggitt has linked the correct line to set them up

pohly commented:
Thanks guys, that should get me further. It's late here now and I am out tomorrow, so I'll look at that on Friday. I'll probably be able to figure out how utils.PrivilegedTestPSPClusterRoleBinding works, but if someone has the corresponding .yaml definitions that might save me some time.

msau42 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -367,57 +385,96 @@ func initCSIgcePD(f *framework.Framework, config framework.VolumeTestConfig) csi
-	role := csiControllerRole(cs, config, false)
-	csiControllerRoleBinding(cs, config, false, role, g.controllerServiceAccount)
-	deployGCEPDCSIDriver(cs, config, false /* teardown */, f, g.nodeServiceAccount, g.controllerServiceAccount)
+	cleanup, err := g.f.CreateFromManifest(func(object interface{}) {
```
fyi, I am planning to launch the gce pd csi driver outside of the e2e test, and disable gce pd csi tests from kubernetes/kubernetes, and run it from kubernetes-sigs/gcp-compute-persistent-disk-csi-driver.  So don't spend too much effort changing gce pd.

pohly commented:
One possible difference is that now the hostpath service account is not bound to any of the pre-defined system roles anymore. Before, it was bound to `system:csi-external-provisioner` and `system:csi-external-attacher`. Does that explain the regression?

Is there some other pre-define system role that the service account should be bound to? Binding it to `e2e-test-privileged-psp` doesn't look right, because that looks like something that is only present in the Kubernetes CI setup, right?

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
So this PR should now pass all tests because of commit "e2e: PSP role binding for hostpath" = 
1f6e9da. I still don't understand why that `RoleBinding` wasn't necessary earlier. I'd be more comfortable proposing this PR for merging if someone could explain that.

I've taken out the log capturing (see PR #69986 for that) and cleaned up the code. Except for this mystery around the PSP `RoleBinding`  this is ready for merging or at least some initial review. IMHO improving the code (see the TODO in `create.go`) can wait until later. If this gets merged with that TODO, I'll file an issue and assign it to me.

pohly commented:
Here's another difference that may be relevant for the permission issue: previously, the e2e.test binary created a single hostpath `Pod`, which succeeded. Now the binary creates a `DaemonSet` and two `StatefulSets`, and the controllers run into the permission issue (`FailedCreate: Error creating: pods "..." is forbidden: unable to validate against any pod security policy: []`). My naive expectation is that the controllers inherit the ability to create pods from the user who created the `DaemonSet` resp. `StatefulSet` - but perhaps that's not how it works?

@liggitt 

My current solution is to add the same `RoleBinding` to the hostpath's `ServiceAccount`s as for gcePD. Is it okay to have a `RoleBinding` that references a non-existent role? Not all clusters will have this `e2e-test-privileged-psp`. I can't test this right now, but will do so on Monday.

pohly commented:
Thanks @liggitt. That clarifies why it worked before and confirms that the `RoleBinding` in this PR is indeed needed. Now we just need to be sure that it has no negative consequences on other clusters and then this PR can be merged.

davidz627 commented:
@pohly sorry I wasn't able to reply as I was out. Looks like the difference here is we used to be bringing up the hostpath as a pod, and you have changed it to a daemonset here. I believe since now its not the "test service account" creating the pod, we now have the daemonset controller (using the daemonset service account) to create the pod. So the daemonset service account must have PSP permissions.

As for why we used the util function and not added it directly. It was because some other tests (non-csi) also required this new PSP permissions and so I made a util function that can be shared so I just reused it for the CSI driver tests as well. There is nothing inherent or special about that function.

pohly commented:
I can confirm that having the PSP `RoleBinding` without the corresponding `ClusterRole` works. As far as I'm concerned, this PR is ready for merging.

@davidz627 @msau42

msau42 commented on test/e2e/framework/framework.go:
```diff
@@ -229,6 +235,10 @@ func (f *Framework) BeforeEach() {
+		f.UniqueName = f.Namespace.GetName()
+	} else {
+		// not guaranteed to be unique, but very likely
+		f.UniqueName = fmt.Sprintf("%s-%08x", f.BaseName, rand.Int31())
```
When is namespace creation not enabled?  And if so, what namespace is used? We have lots of tests that depend on namespace name for generating unique names.

pohly replied:
The framework supports it. Tests explicitly have to ask for it. I don't know which tests make use of that option.

msau42 replied:
Can we just use Namespace.Name and not support parallel drivers tests when it's not set?  At least in our official CI, it creates the namespaces per test and we have many other test cases depending on that behavior.

pohly replied:
We can definitely rely on Namespace.Name and do that in our tests already.

I added this fallback code merely for those tests that might have been written differently. I didn't want to leave f.UniqueName empty in that case, when it isn't that hard to come up with something that serves the same purpose.

msau42 commented on test/e2e/framework/framework.go.orig:
```diff
@@ -0,0 +1,896 @@
+/*
```
should this file be included?

pohly replied:
Urgh. No, thanks for pointing it out. I'll remove it.

pohly replied:
Removed and rebased...

msau42 commented on test/e2e/framework/create.go:
```diff
@@ -0,0 +1,528 @@
+		// which contains all role names that are defined cluster-wide before the test starts?
+		// All those names are excempt from renaming. That list could be populated by querying
+		// and get extended by tests.
+		if item.Name != "e2e-test-privileged-psp" {
```
Or create one per driver?

pohly replied:
I don't know how `e2e-test-privileged-psp` gets set up. It's not done by the `e2e.test` binary at the moment. My understanding was that it is something that has to be configured in the cluster before starting the test.

msau42 commented on test/e2e/framework/create.go:
```diff
@@ -0,0 +1,528 @@
+	Create(f *Framework) (func() error, error)
+}
+
+var Factories = map[What]ItemFactory{
```
This model becomes challenging if we ever have to support CRDs in the future.

pohly replied:
Can you elaborate further what the problem would be? I'm not familiar with CRDs, but looking at https://github.com/kubernetes/csi-api/blob/master/pkg/crd/crd.go and https://github.com/kubernetes/csi-api/blob/master/pkg/crd/testdata/csidriver.yaml, a CRD looks like normal API objects that can be populate using this approach from a .yaml file.

pohly replied:
We don't have to use this approach for CRDs, or we do the patching inside the test which knowns about the CRD. I'm pretty sure that some solution can be found.

msau42 reviewed (commented):
Can we just deploy the drivers once at the very beginning of the suite, instead of once per test case? Then we don't need the extra complexity of having to patch unique names everywhere.

pohly commented:
> Can we just deploy the drivers once at the very beginning of the suite, instead of once per test case?

That has it's own drawbacks:
* the framework doesn't currently allow tests to register global setup code; this would have to be added first
* either *every* invocation, regardless whether it actually runs the tests or not, would need to install and deinstall the drivers (not acceptable when running another test via `-ginkgo.focus`), or we have to somehow coordinate on-demand installation when a test starts (even more complex than the patching, considering that tests might run in different processes in parallel)
* we loose test isolation
* some tests (see `CSI attach test using HostPath driver`) do need to have their own driver under a unique name, otherwise we have to keep running them serially

msau42 commented:
I think it would be ideal to be able to mimic a real production cluster as close as possible, in which case the driver is only deployed once.  Yes, we lose test isolation, however, we can potentially uncover longevity issues in the driver, which is a more realistic workload.  I agree it is currently a challenge to implement this in the current test framework.

The skipAttach tests are not really meant to be tested against a real driver, I am fine special casing renaming logic for those test cases.

pohly commented:
Michelle Au <notifications@github.com> writes:
> I think it would be ideal to be able to mimic a real production
> cluster as close as possible, in which case the driver is only
> deployed once.

So under which name and namespace would you deploy the drivers?

It can't be under the namespace of the current test. We also can't use
the names from the "upstream" .yaml files, because we shouldn't
overwrite something that the developer might have installed manually on
the cluster.

I suspect that renaming in some way (manually or automatically) will be
needed also when deploying only once.

> Yes, we lose test isolation, however, we can potentially uncover
> longevity issues in the driver, which is a more realistic workload.

I agree, that's a good point.

> agree it is currently a challenge to implement this in the current
> test framework.
>
> The skipAttach tests are not really meant to be tested against a real
> driver, I am fine special casing renaming logic for those test cases.

So how about this proposal:
- we merge this PR, thus also unblocking all other work that is
  currently in flight regarding the volume tests
- once someone has a good idea about how to deploy and remove drivers
  once per test suite run and the time to implement that, we
  switch to global deployment for the normal tests and keep per-test
  deployment for skipAttach

msau42 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -367,57 +369,96 @@ func initCSIgcePD(f *framework.Framework, config framework.VolumeTestConfig) csi
+			container := &containers[i]
+			for e := range container.Args {
+				// Inject test-specific provider name into paths (--kubelet-registration-path).
+				container.Args[e] = strings.Replace(container.Args[e], "/"+provisioner+"/", "/"+provisioner+"-"+f.UniqueName+"/", 1)
```
Is this a duplicate of L445?

msau42 replied:
The comment said it's updating kubelet registration path, but the code looks like it's updating the provisioner argument

pohly replied:
The code does what the comment says:
`--kubelet-registration-path=/var/lib/kubelet/plugins/csi-hostpath/csi.sock` becomes `--kubelet-registration-path=/var/lib/kubelet/plugins/csi-hostpath-e2e-tests-csi-volumes-q7vzj/csi.sock`.

I'll give that as full example in the comment.

The comment from above for `--drivername` also applies here: this can work for multiple different drivers because they are likely to follow this convention (or even have to, in this case?).

pohly replied:
Isn't "provisioner" and "driver name" synonymous? Driver registrar expects a `--provisioner` parameter, while hostpath takes a `--drivername`, and both are the same string as far as I understand it.

Please have a look at the revised code (https://github.com/kubernetes/kubernetes/pull/69868/files#diff-10be18decd977f7f8fe7c0c02e2cbf05), perhaps it is clearer there. If not, please let me know how to improve it.

msau42 replied:
There should not be a separate provisioner name.  IMO it was a mistake to make provisioner name configurable.  It should really be equal to the driver name.  But I think overall, we should be referring to driver name, not provisioner name.  Not all drivers may have a provisioner.

msau42 replied:
(I'm not seeing a provisioner parameter do the driver-registrar)

pohly replied:
It takes `--kubelet-registration-path` which happens to get a path which contains "the string that is sometimes the provider name and sometimes the driver name". If it's better to call that string "driver name" then I'll change the code accordingly.

pohly replied:
Done, and added comments explaining that the provisioner name is expected to be the same.

msau42 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -367,57 +369,96 @@ func initCSIgcePD(f *framework.Framework, config framework.VolumeTestConfig) csi
+			case driverContainerName:
+				container.Args = append(container.Args, "--drivername="+provisioner+"-"+f.UniqueName)
+			case provisionerContainerName:
+				container.Args = append(container.Args, "--provisioner="+provisioner+"-"+f.UniqueName)
```
Once https://github.com/kubernetes-csi/external-provisioner/pull/142 makes it into 0.4, then this is no longer needed.

pohly replied:
The new function explicitly allows the container names to be empty, in which case the argument will not be added.

msau42 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -367,57 +369,96 @@ func initCSIgcePD(f *framework.Framework, config framework.VolumeTestConfig) csi
+
+// TODO (?): the storage.csi.image.version and storage.csi.image.registry
+// settings are ignored. We could patch the image definitions.
+// TODO: way too many parameters...
```
I wonder if the selector and labels also need to be patched?
https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/deploy/kubernetes/stable/controller.yaml#L8

msau42 replied:
And the CSI_ADDRESS? https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/deploy/kubernetes/stable/node.yaml#L26

pohly replied:
Selector and labels do not need to be patched because they are relative to the namespace, which is unique to the test. Not renaming when it isn't needed makes debugging easier.

`CSI_ADDRESS` doesn't need to be patched in the manifest used by the E2E testing because it uses `/csi/csi.sock` as fixed path:
https://github.com/kubernetes/kubernetes/blob/689df2010da51b19fec2cdfe3297533cb35c90bd/test/e2e/testing-manifests/storage-csi/gce-pd/node_ds.yaml#L21

IMHO using that full path in `node.yaml` is unnecessarily complex. But perhaps I'm just not aware of some advantage that the full path has?

Apropos that line, what's the advantage of defining an env variable with a fixed value and then referencing that value? That's again something were I don't see the advantage over just `--csi-address=/csi/csi.sock`.

msau42 commented on test/e2e/testing-manifests/storage-csi/hostpath/usage/csi-pvc.yaml:
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
```
do we need manifests for test pvc and pods?

pohly replied:
That could be removed. I kept it for the sake of completeness ("this directory is a copy of..."), but the tests themselves don't need them.

pohly replied:
Done.

msau42 commented:
cc @kubernetes/sig-testing-pr-reviews for thoughts on this approach to be able to create unique objects from manifests

pohly commented on test/e2e/framework/create.go:
```diff
@@ -0,0 +1,528 @@
+			return errors.Wrapf(err, "decode %+v", what)
+		}
+
+		f.PatchItem(object)
```
Yes, patching can be made optional. I'll have a look at whether load/patch/create can be decoupled.

pohly replied:
Done. Loading, patching, creating are now three different functions, with one utility function that roles all into one.

pohly commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -367,57 +369,96 @@ func initCSIgcePD(f *framework.Framework, config framework.VolumeTestConfig) csi
+			// value.
+			switch container.Name {
+			case driverContainerName:
+				container.Args = append(container.Args, "--drivername="+provisioner+"-"+f.UniqueName)
```
Hmm, the gcePD test passed, so the driver must have accepted `--drivername`:
```
I1022 21:12:26.383]             "image": "gcr.io/google-containers/volume-csi/gcp-compute-persistent-disk-csi-driver:v0.1.0.alpha",
I1022 21:12:26.383]             "args": [
I1022 21:12:26.383]               "--v=5",
I1022 21:12:26.383]               "--endpoint=$(CSI_ENDPOINT)",
I1022 21:12:26.383]               "--nodeid=$(KUBE_NODE_NAME)",
I1022 21:12:26.384]               "--drivername=com.google.csi.gcepd-e2e-tests-csi-volumes-w9hsx"
I1022 21:12:26.384]             ],
```
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/69868/pull-kubernetes-e2e-gce/56688/build-log.txt

The patch function in `csi_volumes.go` is driver-specific, it just happens that it intentionally was written so that it works for both hostPath and gcePD. If some other driver doesn't accept the `--drivername` parameter, then it has to use its own patch function or nil, in which case the driver name won't get changed.

Let me add some more documentation to that function to clarify that.

pohly replied:
Done, and I enhanced the configuration of that function and moved it to the utils package. There it will be easier to import into out-of-tree CSI driver test suites.

msau42 replied:
Ah the latest stable version of the PD driver (referenced above) doesn't support it anymore. So it will have to be changed when we update the PD specs

pohly replied:
Hmm, so the latest PD driver will have to be used without driver renaming? Can it be guaranteed that there won't be conflicts between different tests without sorting out the "install driver only once" problem first?

msau42 replied:
cc @davidz627 can you confirm driver name is no longer configurable?  we may need to add it back to allow for this testing model.

For out of tree GCE PD testing, we plan to launch the driver as part of cluster setup and not during the e2e test.

davidz627 replied:
driver name is no longer configurable (https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/129). This was done because we need a stable well-known driver name for migration. Is there no other way to support testing?

pohly commented:
@msau42 updated code pushed, PTAL.

pohly commented:
The hostpath test worked for me locally, but failed in the gce job :-/

pohly committed:
e2e: simplify gcePD .yaml files

Setting command line arguments via env variables that are not needed
by the binaries is just unnecessarily complex. The driver renaming
code in the E2E manifest PR would have to be made more complex to deal
with such a deployment. It is easier for that code and humans who look
at the .yaml file to remove the indirection.

davidz627 commented:
> It was only the gcePD test which failed, the failure just happened to be listed twice.
> 
> The reason for the failure after the rebase is that a recent change made the .yaml file more complex:
> 
> ```
>               "--v=5",
>               "--csi-address=$(ADDRESS)",
>               "--kubelet-registration-path=$(DRIVER_REG_SOCK_PATH)"
>             ],
>             "env": [
>               {
>                 "name": "ADDRESS",
>                 "value": "/csi/csi.sock"
>               },
> ```
> @davidz627 neither @msau42 nor I could think of a reason why it has to be that way, so I simplified the .yaml files again, see [a2ab15f](https://github.com/kubernetes/kubernetes/commit/a2ab15fd4cdf3e85e2cc3b6d61eb6042ed6e3dee). For the sake of consistency and simplicity I did the same for `--csi-address`. Is that okay with you? Can you make the same change in the external repos, to minimize differences?

Could you clarify what you mean by "made more complex" - what exactly changed that was the problem? I don't usually make changes just for the sake of complexity, I'd like to get to the bottom of this and make sure that we're not reverting something I did to fix something else. It is possible I did not document/comment my changes well enough.

msau42 commented:
The main question was why do we need to set an environment variable to a hardcoded value?  Can we simplify it by hardcoding the value directly in the parameter?

davidz627 commented:
> The main question was why do we need to set an environment variable to a hardcoded value? Can we simplify it by hardcoding the value directly in the parameter?

I see, I dont believe that was a recent change. That was from sample deployments a long time ago. I am open to changing it

pohly commented:
/hold

In https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/69868/pull-kubernetes-e2e-gce/57153/ some pods fails to start, but the log only says:
```
Oct 25 19:30:16.148: INFO: At 2018-10-25 19:25:22 +0000 UTC - event for csi-gce-node-5gbck: {kubelet e2e-57153-674b9-minion-group-b3lj} BackOff: Back-off restarting failed container
```

Let's try temporarily with the podlogs PR merged into this one...

pohly commented:
David Zhu <notifications@github.com> writes:
> driver name is no longer configurable
> (https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/129). This
> was done because we need a stable well-known driver name for
> migration. Is there no other way to support testing?

There's currently no way to install drivers from inside the test suite
only once. This is a problem if there is more than one test that needs
the driver.

If this is something that the gcePD driver depends on, then
you and @msau will have to find a way to achieve that, because I don't
have time for it.

pohly commented:
David Zhu <notifications@github.com> writes:

>> The main question was why do we need to set an environment variable to a hardcoded value? Can we simplify it by hardcoding the value directly in the parameter?
>
> I see, I dont believe that was a recent change.

Indeed. The entire --kubelet-registration-path options seems to be new
in Kubernetes master now, that's why I hadn't seen it before the rebase.

msau42 commented:
@pohly to unblock you, I think we should just mark gcePD tests as [Serial] and not do patching of the deployments.

msau42 commented:
@davidz627 to test migration, we'll need to find a way to auto-install PD driver on cluster bringup

pohly committed:
e2e CSI: bump all containers to v0.4.1

pohly commented:
Michelle Au <notifications@github.com> writes:

> @pohly to unblock you, I think we should just mark gcePD tests as [Serial] and not do patching of the deployments.

This is a fallback, in case that renaming is the root cause of the
failure. Right now I am testing with everything updated to 0.4.1,
perhaps that'll work better.

Is there more than one test with gcePD? The manual deployment wasn't
doing renaming either, which apparently worked - or was that just good
luck?

msau42 commented:
There is only one test right now so you may not see issues, however @mkimuram will be adding more once he merges CSI tests with in-tree tests

msau42 commented:
I think if multiple PD drivers are started, then you might see funny issues with multiple sidecars competing with each other, like the external provisioner

pohly commented:
@msau let me come back to the offer of not renaming the gcePD driver. I had that working before the driver update to gcp-compute-persistent-disk-csi-driver:v0.1.0.alpha, but now I can't get it to work again and my possibilities to debug that are rather limited and slow.

Therefore the latest incarnation of this PR deploys the gcePD driver without renaming, as in current Kubernetes master. If that works, please merge. I'm not going to wait for clean test results myself anymore today, though.

I left in the version update - killing two birds with one stone, basically. Otherwise we would have two conflicting PRs.

/hold cancel

msau42 commented on test/e2e/testing-manifests/storage-csi/gce-pd/node_ds.yaml:
```diff
@@ -16,7 +16,7 @@ spec:
         - name: csi-driver-registrar
           imagePullPolicy: Always
-          image: quay.io/k8scsi/driver-registrar:v0.3.0
+          image: quay.io/k8scsi/driver-registrar:v0.4.1
```
Hm I wonder if the alpha pd driver is going to have issues with the latest sidecars.

Also, unfortunately we haven't published public images of the beta gce pd driver yet.  So I think it's best to just leave PD driver image versions the way it was before, and we will update the versions when we're ready.

Also I think we need to add the [Serial] tag to pd so that tests won't be run it parallel.

pohly replied:
The test has passed, so it looks like the alpha driver is compatible with the latest sidecar drivers. Do you want to revert that part nonetheless?

Please lets add the [Serial] tag as part of PR #68025  /cc @mkimuram 

It has been working so far and this PR doesn't make it worse. From a practical perspective,
I don't see a good way to add it to the current test because of the way how it loops over the different drivers, and that part will be changed anyway:
```
	for driverName, initCSIDriver := range csiTestDrivers {
		curDriverName := driverName
		curInitCSIDriver := initCSIDriver

		Context(fmt.Sprintf("CSI plugin test using CSI driver: %s", curDriverName), func() {
```

msau42 replied:
If its passing then it's fine to leave it.  It's not a configuration that's officially been tested/supported by us, but I will update it later once we get the latest driver image published.

pohly replied:
@msau42 so is this PR ready for merging now?

pohly commented:
/test pull-kubernetes-integration

pohly commented:
/test pull-kubernetes-integration

msau42 commented on test/e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-attacher.yaml:
```diff
@@ -0,0 +1,48 @@
+          image: quay.io/k8scsi/csi-attacher:v0.4.0
+          args:
+            - --v=5
+            - --csi-address=$(ADDRESS)
```
do you want to simplify the hostpath specs too?

AishSundar commented:
@pohly is this planned to be merged in 1.13 timeframe? if so please give this a priority and milestone label accordingly. @nikopen in case you need him to apply the milestone label

pohly commented:
Yes, this is for 1.13.

/milestone v1.13
/priority important-soon

pohly commented:
@nikopen can you please add the v1.13 milestone label to this PR?

nikopen commented:
/milestone v1.13

jsafrane commented on test/e2e/framework/create.go:
```diff
@@ -0,0 +1,569 @@
+			Failf("reading manifest file: %v", err)
+		}
+
+		// Split at the "---" separator before working on
```
Would it make sense not to support yaml files with `---`, i.e return error here? It will increase number of yaml files, but it would make more obvious what objects are created. Maybe a test could apply whole directory, something like `kubectl apply -f test/e2e/testing-manifests/storage-csi/hostpath/hostpath`, just using bindata.go "directory" traversing.

jsafrane replied:
ack

msau42 commented:
/lgtm

jsafrane commented:
/assign @saad-ali 
for approval

saad-ali commented:
/approve

gnufied commented on test/e2e/framework/create.go:
```diff
@@ -0,0 +1,569 @@
+// - only the latest stable API version for each item is supported
+func (f *Framework) PatchItems(items ...interface{}) error {
+	for _, item := range items {
+		Logf("patching original content of %T:\n%s", item, PrettyPrint(item))
```
Printing all of the YAMLs is far too noisy - https://github.com/kubernetes/kubernetes/issues/70448
